app/models/sample.rb

```diff
@@ -530,8 +530,15 @@ def metadatum_add_or_update(key, val)
       m = Metadatum.new
       m.key = key
       m.sample = self
-      # Fail if MetadataField doesn't exist
-      m.metadata_field = MetadataField.find_by!(name: key.to_s)
+      if MetadataField.find_by(name: key.to_s)
+        m.metadata_field = MetadataField.find_by(name: key.to_s)
+      elsif MetadataField.find_by(display_name: key.to_s)
+        m.metadata_field = MetadataField.find_by(display_name: key.to_s)
+        # If matched based on display_name, set key to regular name
+        m.key = m.metadata_field.name
+      else
+        raise ActiveRecord::RecordNotFound("No matching field for #{key}")
+      end
```

Comment:
There's some duplication between here and the code in `set_validated_values`. Would it make sense to just set the metadata_field in `set_validated_values` and do `errors.add(...)` if a `metadata_field` doesn't exist? And remove this part entirely.

I guess we eventually want to remove `key` from the Metadatum sample, so maybe it does make sense to do it this way. In which case, we should add a `TODO` to remove the code in `set_validated_values`. Eventually, `metadata_field` MUST be set for Metadatum prior to saving, since there will no longer be a key.

Reply:
@MarkAZhang I think it's good set metadata_field at the point Metadatum.new is called and remove key eventually, so I reworked this a little and removed that from set_validated_values. I checked in staging and prod that all current Metadata entries have a metadata_field.